### PR TITLE
refactoring code to support query level metrics

### PR DESCRIPTION
**Pull request by uporwal**

### CODE REVIEW

File: evaluation_script/main.py [hunk 1]
```diff
@@ -1,5 +1,46 @@
...
+from utils import get_file_extension
+from utils import open_file
+from utils import calculate_precision
+from utils import calculate_recall
+from utils import calculate_fpr
+from utils import calculate_f1
+from utils import BaseMetrics
+from utils import Metrics
+
+query_level_base_metrics = {}
```
Comment (jondegenhardt) [inline evaluation_script/main.py:10]: A bit cleaner would be to create a new instance each of these three dictionaries with each invocation of `evaluate`. The advantage being that evaluation could be called multiple times in the same python invocation. At the moment though multiple calls to evaluate will get incorrect results because these are not being reset. In fact, the unit tests may fail due to this.

The simple thing to do without doing further restructuring is to clear these at the start of the evaluate function, similar to how the `tp`, `fp`, etc variables are getting set to zero.

Comment (jondegenhardt) [reply evaluation_script/main.py:10]: Fixed in latest version.

File: evaluation_script/main.py [hunk 2]
```diff
@@ -1,5 +1,48 @@
...
+"""
+def calculate_query_level_metrics():
+    qa_precision = 0
+    qa_recall = 0
+    qa_fpr = 0
+    qa_f1 = 0
+
+    for query_id in query_level_base_metrics:
+        base_metrics = query_level_base_metrics[query_id]
+        if not are_all_base_metrics_zero(base_metrics.tp, base_metrics.tn, base_metrics.fp, base_metrics.fn):
```
Comment (jondegenhardt) [inline evaluation_script/main.py]: Should the test for metrics being zero occur inside of the `calculate_xyz` routines?

File: evaluation_script/main.py [hunk 3]
```diff
@@ -1,5 +1,48 @@
...
+            query_level_metrics[query_id].f1 = calculate_f1(base_metrics.tp, base_metrics.fp, base_metrics.fn)
+
+    for query_id in query_level_base_metrics:
+        qa_precision = qa_precision + query_level_metrics[query_id].precision
+        qa_recall = qa_recall + query_level_metrics[query_id].recall
+        qa_fpr = qa_fpr + query_level_metrics[query_id].fpr
+        qa_f1 = qa_f1 + query_level_metrics[query_id].f1
+
+    total_queries = len(query_level_base_metrics.keys())
+    if total_queries > 0:
```
Comment (jondegenhardt) [inline evaluation_script/main.py]: `total_queries > 0` is handling an unexpected edge case. Would be clearer to put it at the start of processing, right after the declarations of `qa_xzy` variables.

Comment (uporwal) [reply evaluation_script/main.py]: Will fix this.

File: evaluation_script/main.py [hunk 4]
```diff
@@ -53,32 +87,39 @@ def calculate_base_metrics(infile, truth):
...
-                            else:
-                                tn = tn + 1
-                        else:
-                            if truth[(query_id, doc_id)] == '1':
-                                fn = fn + 1
+            if doc_id in documents_with_ground_truth:
+                for i in range(1, length):
+                    query_id = index[i]
+                    if (query_id, doc_id) in truth:
+                        if (query_id, doc_id) not in predicted_keys:
```
Comment (jondegenhardt) [inline evaluation_script/main.py:95]: The `if (query_id, doc_id) not in predicted_keys:` test is protecting against both duplicate rows (duplicate docids) and duplicate columns (duplicate query ids), is that correct?

File: evaluation_script/main.py (hunk 2 above)
Comment (uporwal) [reply evaluation_script/main.py]: No, because `calculate_xyz` will check only for that particular metric but I need to check if they are all `0` at once. If some of them are `0`, that is ok. I need this handling only when all of these become `0`. If I wont put it then `F1` gets a value of `1` because if `tp`, `fp` and `fn` are all `0`, precision and recall get the value of `1`. I dont want the `F1` to be `1` in such scenario. This check helps me prevent that case.

File: evaluation_script/main.py (hunk 4 above)
Comment (uporwal) [reply evaluation_script/main.py:95]: Yes.

File: evaluation_script/main.py [hunk 5]
```diff
@@ -186,28 +196,36 @@ def evaluate(test_annotation_file, user_submission_file, phase_codename, **kwarg
...
-            precision = calculate_precision(tp, fp)
-            recall = calculate_recall(tp, fn)
-            fpr = calculate_recall(fp, tn)
-            f1 = calculate_f1(precision, recall)
-        else:
-            precision = 0
-            recall = 0
-            fpr = 0
-            f1 = 0
+            if not are_all_base_metrics_zero(tp, tn, fp, fn):
```
Comment (jondegenhardt) [inline evaluation_script/main.py]: Same question as line 26 - Why isn't the `are_all_base_metrics_zero` test occurring inside the metric calculations?

File: evaluation_script/main.py [hunk 6]
```diff
@@ -186,28 +196,36 @@ def evaluate(test_annotation_file, user_submission_file, phase_codename, **kwarg
...
-                    "recall": recall,
-                    "f1": f1,
-                    "tpr": recall,
-                    "fpr": fpr,
+                    "global_precision": precision,
+                    "global_recall": recall,
+                    "global_f1": f1,
+                    "global_tpr": recall,
+                    "global_fpr": fpr,
+                    "averaged_precision": qa_precision,
```
Comment (jondegenhardt) [inline evaluation_script/main.py]: Use "average" rather than "averaged". It's the more common term.

File: tests.py [hunk 7]
```diff
@@ -26,66 +35,135 @@ def test_g1_p1(self):
...
 
     def test_g2_p2a(self):
-        expected = {'precision': 0.5, 'recall': 1.0, 'f1': 0.6666666666666666, 'tpr': 1.0, 'fpr': 1.0}
+        expected = {'global_precision': 0.5,
+                    'global_recall': 1.0,
+                    'global_f1': 0.6666666666666666,
+                    'global_tpr': 1.0,
+                    'global_fpr': 1.0,
+                    'averaged_precision': 0.4375,
+                    'averaged_recall': 0.875,
```
Comment (jondegenhardt) [inline tests.py]: `average_recall == 0.875` suggests an edge-case error in our calculations. The `predictions_set2a.tsv` file predicts all items are relevant. So how could the results be anything but `1.0`? Likely because `ground_truth_set2.tsv has:
* One query has zero judgements
* One query has only irrelevant judgements

Probably the latter. I'll take a look at calculations, but perhaps recall is being set to zero if there are zero relevant docs. It should probably be one.

Also though, if a query has zero judged documents, the query should not be included when counting the average.

**Update:** I suspect the problem is the handling of queries with no judgements. What might be cleanest is to keep a `set` of queries having judgements, or perhaps a `dictionary` mapping queries to the number of judgements they have. (The latter may be useful for other purposes.) Then in `calculate_query_level_metrics` only score queries that have at least one judgement.

File: evaluation_script/main.py (hunk 5 above)
Comment (uporwal) [reply evaluation_script/main.py]: Same response as line 26. Let me know if that makes sense.

File: evaluation_script/main.py (hunk 6 above)
Comment (uporwal) [reply evaluation_script/main.py]: will do

File: tests.py (hunk 7 above)
Comment (uporwal) [reply tests.py]: It is because of the first option. I am counting all queries and one query has all base metrics as `0` hence precision, recall, f1 are also `0`. Therefore, we have an average of 7/8 as `0.875`. I'll exclude queries with zero judged documents from metrics calculation. That would resolve this.

File: tests.py [hunk 8]
```diff
@@ -26,66 +35,135 @@ def test_g1_p1(self):
...
 
     def test_g2_p2b(self):
-        expected = {'precision': 1, 'recall': 0.0, 'f1': 0.0, 'tpr': 0.0, 'fpr': 0.0}
+        expected = {'global_precision': 1,
+                    'global_recall': 0.0,
+                    'global_f1': 0.0,
+                    'global_tpr': 0.0,
+                    'global_fpr': 0.0,
+                    'average_precision': 1.0,
+                    'average_recall': 0.142857143,
```
Comment (jondegenhardt) [inline tests.py:61]: Hm... Shows a logical problem with the query metrics. Oh well. It's correct mathematically, not much to be done.

In this case the predictions are saying that nothing is relevant. However, for queries with no relevant results this has a recall at `1.0`, which get averaged in. This could actually occur in the data challenge, as I didn't take any steps to ensure that every query has at least one relevant judgement in the evaluation set, and a few queries have only a small number of relevant results.

File: tests.py (hunk 7 above)
Comment (jondegenhardt) [reply tests.py]: Fixed in revision

File: evaluation_script/main.py (hunk 6 above)
Comment (jondegenhardt) [reply evaluation_script/main.py]: Fixed in revision

File: evaluation_script/main.py (hunk 3 above)
Comment (jondegenhardt) [reply evaluation_script/main.py]: Fixed in revision
